reserved_keyword_checker/reserved_keyword_checker.py

```diff
@@ -0,0 +1,292 @@
+"""
+
+import inspect
+import io
+import re
+import os
+import sys
+
+import click
+import django
+import yaml
+from django.db import models
+
+
+class Violation(object):
+    """
+    A Django model field name that is in conclict with a defined list of reserved keywords
+    """
+
+    def __init__(self, model, field_name, system, override=False):
+        self.model = model
+        self.field = field_name
+        self.system = system
+        self.override = override
+
+    def __str__(self):
+        return "{} conflict in {}:{}:{}.{}".format(
+            self.system,
+            self.model._meta.app_label,
+            self.module_name,
+            self.model._meta.concrete_model.__name__,
+            self.field
+        )
+
+    def report_string(self):
+        """
+        Generate a comma-separated string used for creating a report of the reserved
+        keywords that were found. Output should be returned with the following fields:
+        - system: the system/tool that has the restriction that triggered the violation
+        - app source: was the violation detected in a locally defined or third party installed
+          application
+        - app_name: name of the Django app that contains the violation
+        - module_name: the file path of the Python module containing the violation
+        - model_name: name of the Django model that contains the violation
+        - field: the actual offending field name that triggered the violation
+        - source: is this module/model where the offending field was defined, or was it
+          inherited from another model.
+        """
+        model_name = self.model._meta.concrete_model.__name__
+        app_name = self.model._meta.app_label
+        if self.local_app:
+            app_source = "Local"
+        else:
+            app_source = "3rd party"
+        if not self.inherited:
+            source = "Defined here"
+        else:
+            source = "Inherited"
+        return "{},{},{},{},{},{},{}".format(
+            self.system, app_source, app_name, self.module_name, model_name, self.field, source
+        )
+
+    @property
+    def module_name(self):
+        """
+        The path to the module containing the reserved keyword violation
+        """
+        module_string = re.sub(r'\.', '/', self.model._meta.concrete_model.__module__)
+        return "{}.py".format(module_string)
+
+    @property
+    def inherited(self):
+        """
+        Return whether or not this violation is defined in a parent model of the model
+        in which it was found
+        """
+        return self.field not in [f.name for f in self.model._meta.local_fields]
+    
+    @property
+    def local_app(self):
+        """
+        Return whether or not this violation was found in a local source file or a third
+        party installed app
+        """
+        return os.path.isfile(self.module_name)
```

Comment:
I'm not sure about this line. I know we did this differently in code-annotations, and it was kind of challenging. Is there a source somewhere that describes how this works?

Reply:
It seems like we're missing a check for "site-packages" or "dist-packages" in the path string:

> On Posix systems according to our testing, non-local packages get installed to paths containing either "site-packages" or "dist-packages".

https://github.com/edx/code-annotations/blob/master/code_annotations/find_django.py#L333

Reply:
I have added tests for this and they pass. I might be missing something, but can we just consider modules that exist in the source code to be local and all others to be non-local? As it is currently stands, this just checks if the module path exists in the directory that the scan is being run in. Module path as defined above, would never include a virtualenv structure in the path and therefore could never be found by this method.

Reply:
Oh, so you're asserting that the module is local by constructing a relative path and searching from the $PWD.  Whereas, we had previously asserted the opposite by constructing an absolute path and looking for evidence of a library directory.  I think I actually like your method better, assuming it works in every case.

However, would it still work if the virtualenv was created under the PWD?  Anyway, that's not how most CI creates virtualenvs, so this isn't a blocker.

Reply:
What if it's not run from the assumed PWD?